pyscraper/new_hansard.py

```diff
@@ -0,0 +1,1425 @@
+            'http://www.parliament.uk/lords/hansard/print'
+        ),
+        'pbc': (
+            '//ns:System[@type="Debate"]',
+            'http://www.parliament.uk/commons/hansard/print'
+        ),
+    }
+
+    oral_headings = [
+        'hs_3OralAnswers'
+    ]
+    major_headings = [
+        'hs_6bDepartment',
+        'hs_6bBigBoldHdg',
+        'hs_2cBillTitle',
+        'hs_2cUrgentQuestion',
+        'hs_3cMainHdg',
+        'hs_2BusinessWODebate',
+        'hs_2cStatement',
+        'hs_2BillTitle',
+        'hs_6bBillTitle',
+        'hs_6bBusinessB4Questions',
+        'hs_6bPrivateBusiness',
+        'hs_6bRoyalAssent',
+        'hs_6bBillsPresented', # FIXME should grab text of following tag
+        'hs_6fCntrItalHdg',
+        'hs_2cSO24Application',
+    ]
+    minor_headings = [
+        'hs_8Question',
+        'hs_6bFormalmotion',
+        'hs_7SmCapsHdg',
+        'hs_7Bill',
+        'hs_6bcBigBoldHdg',
+    ]
+    generic_headings = [
+        'hs_2cDebatedMotion',
+        'hs_2cGenericHdg',
+        'hs_2GenericHdg',
+    ]
+    whall_headings = [
+        'hs_2DebBill',
+        'hs_2cWestHallDebate',
+        'hs_2WestHallDebate'
+    ]
+    paras = [
+        'hs_Para',
+        'hs_AmendmentLevel1',
+        'hs_AmendmentLevel2',
+        'hs_AmendmentLevel3',
+        'hs_AmendmentLevel4',
+        'hs_8Clause',
+        'hs_newline10',
+        'hs_Question',
+    ]
+    indents = [
+        'hs_ParaIndent',
+        'hs_AmendmentLevel0'
+    ]
+    ignored_tags = [
+        'hs_6bPetitions',
+        'hs_76fChair',
+        'hs_3MainHdg',
+        'hs_Venue'
+    ]
+    root = None
+    ns = ''
+    ns_map = {}
+
+    debate_type = None
+    current_speech = None
+    date = ''
+    rev = 'a'
+    use_pids = True
+    current_col = 0
+    current_speech_num = -1
+    current_speech_part = 1
+    current_time = ''
+    output_heading = False
+    skip_tag = None
+    uc_titles = False
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.debate_type = None
+        self.current_speech = None
+        self.date = ''
+        self.rev = 'a'
+        self.current_col = 0
+        self.current_speech_num = 0
+        self.current_speech_part = 1
+        self.current_time = ''
+        self.root = None
+        self.input_root = None
+        self.output_heading = False
+
+    def get_tag_name_no_ns(self, tag):
+        # remove annoying namespace for brevities sake
+        tag_name = str(tag.tag)
+        tag_name = tag_name.replace(
+            '{{{0}}}'.format(self.ns),
+            ''
+        )
+        return tag_name
+
+    def get_pid(self):
+        if self.current_speech_num == -1:
+            self.current_speech_num = 0
+        return '{0}{1}.{2}/{3}'.format(
+            self.rev,
+            self.current_col,
+            self.current_speech_num,
+            self.current_speech_part
+        )
+
+    def get_speech_id(self):
```

Comment:
Not having this function do the incrementing leaves you open to missing places to increment it. Currently, it looks like parse_oral_heading, parse_division and parse_WHDebate do not increment, leading to duplicate IDs. Conversely it looks like parse_major may increment twice (by calling clear_current_speech and doing it itself); there appears to be some things that pre-increment, and some things that post-increment.